components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,162 @@
+
+                        if (terminal != null) {
+                            createSharedTerminal(terminal)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun getTerminalToolWindowRegisteredEvent(): CompletableFuture<Void> {
+        val completableFuture = CompletableFuture<Void>()
+
+        debug("Waiting for TerminalToolWindow to be registered...")
+        val toolWindowManagerListener =
+                object : ToolWindowManagerListener {
+                    override fun toolWindowsRegistered(
+                            ids: MutableList<String>,
+                            toolWindowManager: ToolWindowManager
+                    ) {
+                        if (ids.contains(TerminalToolWindowFactory.TOOL_WINDOW_ID)) {
+                            debug("TerminalToolWindow got registered!")
+                            completableFuture.complete(null)
+                        }
+                    }
+                }
+
+        project.messageBus
+                .connect()
+                .subscribe(ToolWindowManagerListener.TOPIC, toolWindowManagerListener)
+
+        return completableFuture
+    }
+
+    private fun getSupervisorTasksList(): CompletableFuture<List<TaskStatus>> {
+        val completableFuture = CompletableFuture<List<TaskStatus>>()
+        val taskStatusRequest = TasksStatusRequest.newBuilder().setObserve(true).build()
+        val taskStatusResponseObserver =
+                object : StreamObserver<TasksStatusResponse> {
+                    override fun onNext(response: TasksStatusResponse) {
+                        debug("Received task list: ${response.tasksList}")
+
+                        var hasOpenedAllTasks = true
+
+                        response.tasksList.forEach { task ->
+                            if (task.state === TaskState.opening) {
+                                hasOpenedAllTasks = false
+                            }
+                        }
+
+                        if (hasOpenedAllTasks) {
+                            this.onCompleted()
+                            completableFuture.complete(response.tasksList)
+                        }
+                    }
+
+                    override fun onCompleted() {
+                        debug("Successfully fetched tasks from Supervisor.")
+                    }
+
+                    override fun onError(throwable: Throwable) {
+                        thisLogger()
+                                .error(
+                                        "Got an error while trying to fetch tasks from Supervisor.",
+                                        throwable
+                                )
+                        completableFuture.completeExceptionally(throwable)
+                    }
+                }
+
+        statusServiceStub.tasksStatus(taskStatusRequest, taskStatusResponseObserver)
+
+        return completableFuture
```

Comment:
I will assume there should be something like `while(isAcvtive)` and try catch to handle network errors, there only exception which can allow to exit is cancellation

Reply:
Understood. I'll look more into it soon. Currently `getSupervisorTasksList` is wrapped within a `retry(3)`.

Reply:
Alright, all `retry` have been replaced by `while(isAcvtive)`.